links/src/main/java/juja/microservices/links/repository/impl/LinksRepositoryImpl.java

```diff
@@ -1,18 +1,61 @@
 package juja.microservices.links.repository.impl;
 
 import juja.microservices.links.model.Link;
+import juja.microservices.links.model.SaveLinkRequest;
 import juja.microservices.links.repository.LinksRepository;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.stereotype.Repository;
 
+import javax.inject.Inject;
 import java.util.Set;
 
-/**
- * @author Ivan Shapovalov
- */
+import static org.springframework.data.mongodb.core.query.Criteria.where;
+import static org.springframework.data.mongodb.core.query.Query.query;
+
+@Slf4j
+@Repository
 public class LinksRepositoryImpl implements LinksRepository {
+    @Inject
+    private MongoTemplate mongoTemplate;
+
+    @Value("${spring.data.mongodb.collection}")
+    private String mongoCollectionName;
 
     @Override
-    public Link saveLink() {
-        return null;
+    public Link saveLink(SaveLinkRequest request) {
+        Link link;
+        String URL = request.getURL();
+        String id = getLinkId(URL);
+
+        if (id.isEmpty()) {
+            mongoTemplate.save(request, mongoCollectionName);
+            link = getLinkByURL(URL);
+            if (link == null) {
+                log.error("Failed to save link. '{}'", request);
+            } else {
+                log.info("Successfully saved link. URL: {}, id: [{}]", URL, link.getId());
+            }
+        } else {
+            link = getLinkByURL(URL);
+            log.info("Link already saved. URL: {}, id: [{}]", URL, id);
+        }
+
+        return link;
+    }
```

Comment:
All method could be the same

    @Override
    public Link saveLink(SaveLinkRequest request) {
        Link link =  mongoTemplate.findOne(query(where("url").is(request.getUrl())), Link.class, 
                mongoCollectionName);

        if (link==null) {
            link=new Link(request.getUrl());
            mongoTemplate.save(link, mongoCollectionName);
            log.info("Successfully saved link. Link [{}]", link);
        } else {
            log.info("Link already saved. Link: [{}]", link);
        }

        return link;
    }

Reply:
Or maybe good way is remove all logic into service. Repo have two small methods such as saveLink(Link link) and findLinkByUrl(String url)

Reply:
changed